x-pack/plugins/security_solution/server/endpoint/services/metadata/endpoint_metadata_service.ts

```diff
@@ -0,0 +1,238 @@
+   * for endpoint enrichment (so that users are not required to have superuser role)
+   *
+   * **IMPORTANT: SHOULD BE USED ONLY FOR READ-ONLY ACCESS AND WITH DISCRETION**
+   *
+   * @private
+   */
+  public get DANGEROUS_INTERNAL_SO_CLIENT() {
+    // The INTERNAL SO client must be created during the first time its used. This is because creating it during
+    // instance initialization (in `constructor(){}`) causes the SO Client to be invalid (perhaps because this
+    // instantiation is happening during the plugin's the start phase)
+    if (!this.__DANGEROUS_INTERNAL_SO_CLIENT) {
+      const fakeRequest = ({
+        headers: {},
+        getBasePath: () => '',
+        path: '/',
+        route: { settings: {} },
+        url: { href: {} },
+        raw: { req: { url: '/' } },
+      } as unknown) as KibanaRequest;
+
+      this.__DANGEROUS_INTERNAL_SO_CLIENT = this.savedObjectsStart.getScopedClient(fakeRequest, {
+        excludedWrappers: ['security'],
+      });
+    }
+
+    return this.__DANGEROUS_INTERNAL_SO_CLIENT;
+  }
+
+  /**
+   * Retrieve a single endpoint host metadata. Note that the return endpoint document, if found,
+   * could be associated with a Fleet Agent that is no longer active. If wanting to ensure the
+   * endpoint is associated with an active Fleet Agent, then use `getEnrichedHostMetadata()` instead
+   *
+   * @param esClient Elasticsearch Client (usually scoped to the user's context)
+   * @param endpointId the endpoint id (from `agent.id`)
+   *
+   * @throws
+   */
+  async getHostMetadata(esClient: ElasticsearchClient, endpointId: string): Promise<HostMetadata> {
+    const query = getESQueryHostMetadataByID(endpointId);
+    const queryResult = await esClient.search<HostMetadata>(query).catch(catchAndWrapError);
+    const endpointMetadata = queryResponseToHostResult(queryResult.body).result;
+
+    if (endpointMetadata) {
+      return endpointMetadata;
+    }
+
+    throw new EndpointHostNotFoundError(`Endpoint with id ${endpointId} not found`);
+  }
+
+  /**
+   * Retrieve a single endpoint host metadata along with fleet information
+   *
+   * @param esClient Elasticsearch Client (usually scoped to the user's context)
+   * @param endpointId the endpoint id (from `agent.id`)
+   *
+   * @throws
+   */
+  async getEnrichedHostMetadata(
+    esClient: ElasticsearchClient,
+    endpointId: string
+  ): Promise<HostInfo> {
+    const endpointMetadata = await this.getHostMetadata(esClient, endpointId);
+
+    let fleetAgentId = endpointMetadata.elastic.agent.id;
+    let fleetAgent: Agent | undefined;
+
+    // Get Fleet agent
+    try {
+      if (!fleetAgentId) {
+        fleetAgentId = endpointMetadata.agent.id;
+        this.logger?.warn(`Missing elastic agent id, using host id instead ${fleetAgentId}`);
+      }
+
+      fleetAgent = await this.getFleetAgent(esClient, fleetAgentId);
+    } catch (error) {
+      if (error instanceof FleetAgentNotFoundError) {
```

Comment:
I am not clear why this was done this way in the prior implementation, but I kept the same behaviour. Personally, I would have preferred that we `throw` for NotFound as well, since we always have to assume that Endpoint and agent are always going to be together - but perhaps I'm missing something.

Reply:
What if a user misconfigured something. Or was doing aggressive cleanup without knowing what they were doing and deleted `.fleet-agents`. Or maybe there's a disk (full, corrupt) error, and we can't get the fleet data, but we _do_ have primary Endpoint info. Do we show nothing at all, or do we show the Endpoint info we got (which is 85% of the info anyway).

Reply:
True - all "if" cases. and actually, the one that I think is more common that I just remembered is the one where the agent has been unenrolled, but the endpoint is still being displayed on the list. I actually think that is why this was done that way. 

Will keep it as is. Thanks for the comment. 😄